boot/assets.go

```diff
@@ -50,19 +151,47 @@ type TrustedAssetsInstallObserver struct {
 //
 // Implements gadget.ContentObserver.
 func (o *TrustedAssetsInstallObserver) Observe(op gadget.ContentOperation, affectedStruct *gadget.LaidOutStructure, root, realSource, relativeTarget string) (bool, error) {
-	// TODO:UC20:
-	// steps on write action:
-	// - copy new asset to assets cache
-	// - update modeeenv
+	if affectedStruct.Role != gadget.SystemBoot {
+		// only care about system-boot
+		return true, nil
+	}
+
+	if o.bl == nil {
+		bl, err := bootloader.ForGadget(o.gadgetDir, root, &bootloader.Options{NoSlashBoot: true})
+		if err != nil {
+			return false, fmt.Errorf("cannot find bootloader: %v", err)
+		}
+		tbl, ok := bl.(bootloader.TrustedAssetsBootloader)
+		if !ok {
+			return true, nil
+		}
+		trustedAssets, err := tbl.TrustedAssets()
+		if err != nil {
+			return false, fmt.Errorf("cannot list %q bootloader trusted assets: %v", bl.Name(), err)
+		}
+		o.trustedAssets = trustedAssets
+		o.bl = bl
+	}
+	if len(o.trustedAssets) == 0 || !strutil.ListContains(o.trustedAssets, relativeTarget) {
+		// not one of the trusted assets
+		return true, nil
+	}
+	ta, added, err := o.cache.Add(realSource, o.bl.Name(), filepath.Base(relativeTarget))
+	if err != nil {
+		return false, err
+	}
+	if added {
+		o.trackingAssets = append(o.trackingAssets, ta)
+	}
 	return true, nil
 }
 
-// Seal performs the initial sealing of encryption key to the TPM device
-// available in the system.
-func (o *TrustedAssetsInstallObserver) Seal() error {
-	// TODO:UC20: steps:
-	// - initial seal
-	return nil
+func (o *TrustedAssetsInstallObserver) currentTrustedBootAssetsMap() bootAssetsMap {
+	bm := bootAssetsMap{}
+	for _, tracked := range o.trackingAssets {
+		bm[tracked.name] = append(bm[tracked.name], tracked.hash)
```

Comment:
in the install case we shouldn't have multiple hashes for the same asset, this merits a comment or the code to be written to make one element lists

Reply:
as I said we should add a comment like

// we expect to add exactly one hash per tracked assets